src/include/OSL/rendererservices.h

```diff
@@ -189,6 +206,28 @@ class OSLEXECPUBLIC RendererServices {
                          const EncodedType* arg_types, uint32_t arg_values_size,
                          uint8_t* arg_values);
 
+    /// @brief Builds a free function to provide a value for a given attribute.
+    /// This occurs at shader compile time, not at execution time.
+    /// @param group The shader group currently requesting the attribute.
+    /// @param object_lookup True if an object name was specified, even if the
+    /// value is not known at compile time.
+    /// @param object_name The object name. A null pointer will be provided if
+    /// the value is not specified or it is not known at compile time.
+    /// @param attribute_name The attribute name. A null pointer will be
+    /// provided if the value is not known at compile time.
+    /// @param array_lookup True if the attribute lookup provides an index.
+    /// @param array_index. The array index. A null pointer will be provided if
+    /// the value is not specified or it is not known at compile time.
+    /// @param type The type of the value being requested.
+    /// @param derivatives True if derivatives are also being requested.
+    /// @param spec The built attribute getter. An empty function name is
+    /// interpreted as a missing attribute.
+    virtual void build_attribute_getter(ShaderGroup& group, bool object_lookup,
+                                        ustring* object_name,
+                                        ustring* attribute_name,
+                                        bool array_lookup, int* array_index,
+                                        TypeDesc type, bool derivatives,
+                                        AttributeGetterSpec& spec);
```

Comment:
Sorry, I've got a few minor last-minute concerns about this addition to RendererServices, just want to be sure we're nailing down the parameters before it becomes part of the public API.

    virtual void build_attribute_getter(ShaderGroup& group, bool object_lookup,
                                        ustring* object_name,
                                        ustring* attribute_name,
                                        bool array_lookup, int* array_index,
                                        TypeDesc type, bool derivatives,
                                        AttributeGetterSpec& spec);

What's the difference between object_lookup being false and object_name being null? Is the object_lookup parameter needed? Same question with array_lookup and array_index.

Should any or all of those pointer or reference parameters be const?

Reply:
Oh, I think I see now? The bool can be true and the pointer null in situations where the lookup will provide a value, but it isn't known at compile time? 

If that's the case, then my only concern here is whether those ptrs/refs should be const.

Reply:
OK, let's make sense. 

Please const-ify what makes sense, then I'm ready to merge.

Reply:
I'm not worried about this. The documentation for the function explains it just fine. I do think it should be `const ustring*` to be extra clear that the function is not going to change what it points to. But the rest does not concern me at all.